File: solutions/clojure/027-search-in-rotated-sorted-array.clj
;; Problem 27: Search in Rotated Sorted Array (LeetCode 33)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn search-rotated [nums target]
  (let [arr (vec nums)]
    (loop [lo 0 hi (dec (count arr))]
      (if (> lo hi) -1
        (let [mid (quot (+ lo hi) 2)
              m (arr mid) l (arr lo) h (arr hi)]
          (cond
            (= m target) mid
            (<= l m)
              (if (and (>= target l) (< target m))
                (recur lo (dec mid))
                (recur (inc mid) hi))
            :else
              (if (and (> target m) (<= target h))
                (recur (inc mid) hi)
                (recur lo (dec mid)))))))))

File: solutions/typescript/027-search-in-rotated-sorted-array.ts
/**
 * Problem 27: Search in Rotated Sorted Array (LeetCode 33)
 * Difficulty: Med
 * Language: TypeScript
 */
function search(nums: number[], target: number): number {
  const go = (lo: number, hi: number): number => {
    if (lo > hi) return -1;
    const mid = Math.floor((lo + hi) / 2);
    if (nums[mid] === target) return mid;
    if (nums[lo] <= nums[mid]) {
      return target >= nums[lo] && target < nums[mid]
        ? go(lo, mid - 1)
        : go(mid + 1, hi);
    }
    return target > nums[mid] && target <= nums[hi]
      ? go(mid + 1, hi)
      : go(lo, mid - 1);
  };
  return go(0, nums.length - 1);
}
